### PR TITLE
core:add OPENCV_IPP_MEAN/MINMAX/SUM option to enable IPP optimizations

### PR DESCRIPTION
**Pull request by Kumataro**
fix https://github.com/opencv/opencv/issues/24145
fix https://github.com/opencv/opencv/issues/24146
relates #13085

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/core/CMakeLists.txt [hunk 1]
```diff
@@ -60,6 +60,23 @@ if(CV_TRACE AND HAVE_ITT)
   add_definitions(-DOPENCV_WITH_ITT=1)
 endif()
 
+# https://github.com/opencv/opencv/issues/24145
+ocv_check_environment_variables(OPENCV_IPP_MEAN)
```
Comment (opencv-alalek) [inline modules/core/CMakeLists.txt:64]: Lets guard this block by `if(HAVE_IPP)`

File: modules/core/CMakeLists.txt [hunk 2]
```diff
@@ -60,6 +60,23 @@ if(CV_TRACE AND HAVE_ITT)
...
+option(OPENCV_IPP_MEAN "Enable IPP optimizations for mean (+200Kb in binary size)" OFF)
+if(OPENCV_IPP_MEAN)
+  add_definitions(-DOPENCV_IPP_MEAN=1)
```
Comment (opencv-alalek) [inline modules/core/CMakeLists.txt:67]: It is better to apply definition to the specific file:

```
ocv_append_source_file_compile_definitions(${CMAKE_CURRENT_SOURCE_DIR}/src/mean.dispatch.cpp "OPENCV_IPP_MEAN=1")
```

**Review by opencv-alalek — COMMENTED** [review]
Thank you for contribution!

BTW, It make sense to think about something like `OPENCV_IPP_ENABLE_ALL` in the future.

File: modules/core/CMakeLists.txt (hunk 1 above)
Comment (Kumataro) [reply modules/core/CMakeLists.txt]: I agree with you, thank you! 
And I tried with raspi(arm32, not intel), `OPENCV_IPP_GAUSSIAN_BLUR` is shown in cmake gui. So I added fot it also.

File: modules/core/CMakeLists.txt (hunk 2 above)
Comment (Kumataro) [reply modules/core/CMakeLists.txt]: Thank you for yor review, this is more  better fix. I fix with it.

**Comment by Kumataro** [comment]
Hello, thank you for your review.

>BTW, It make sense to think about something like OPENCV_IPP_ENABLE_ALL in the future.

In cmake, it seems to be a complicated to link multiple options.

To simplify selecting to optimization with ipp, there are 3 ideas.
In Now Idea  A is selected. Is Idea B or C better than A ?

- Idea A) **Disable** OPENCV_IPP_MEAN/MINMAX/SUM/GAUSSIAN_BLUR **(+0MiB)**. People who want to improve performance can enable it in the user option.
- Idea B) **Enable** OPENCV_IPP_MEAN/MINMAX/SUM and **disable** OPENCV_IPP_GAUSSIAN_BLUR **(+1MiB).**.  This method may be a good balance between size and performance.
- Idea C) **Enable** OPENCV_IPP_MEAN/MINMAX/SUM/GAUSSIAN_BLUR **(+9MiB)**. People who want to make the library size smaller can remove it in the user option.

In particular, in an environment where IPP can be used, even if the library size increases by several MiB, there may be almost no effect on operation. In this case, Idea C is better than other idea.

**Comment by opencv-alalek** [comment]
I mean something like this: https://github.com/opencv/opencv/blob/4.8.0/CMakeLists.txt#L218-L225

Idea is using `OPENCV_IPP_ENABLE_ALL` as initial value for these options. User still may override values of options one-by-one.

**Review by opencv-alalek — APPROVED** [review]
Thank you :+1:

**Comment by Kumataro** [comment]
I was over-considered the problem.It is clearly, 

Thank you very much!  I'll try it tomorrow.

File: modules/core/CMakeLists.txt [hunk 3]
```diff
@@ -60,6 +60,26 @@ if(CV_TRACE AND HAVE_ITT)
   add_definitions(-DOPENCV_WITH_ITT=1)
 endif()
 
+# https://github.com/opencv/opencv/issues/24145
+if(HAVE_IPP)
+  OCV_OPTION(OPENCV_IPP_ENABLE_ALL "Force to enable all ipp optimization" OFF)
```
Comment (opencv-alalek) [inline modules/core/CMakeLists.txt:65]: ipp => IPP

---

all => available ?

some implementations are still disabled due to accuracy or performance issues

Comment (Kumataro) [reply modules/core/CMakeLists.txt]: Thank you for your comment! 

I would like to make fix based on the Description of OPENCV_FORCE_3RDPARTY_BUILD.

> OCV_OPTION(OPENCV_IPP_ENABLE_ALL "Enable all OPENCV_IPP_ options at once" OFF)

https://docs.opencv.org/4.x/db/d05/tutorial_config_reference.html

**Pull request merged**